compiler/rustc_trait_selection/src/error_reporting/traits/mod.rs

```diff
@@ -583,3 +607,23 @@ pub fn report_dyn_incompatibility<'tcx>(
 
     err
 }
+
+fn attempt_dyn_to_impl_suggestion(tcx: TyCtxt<'_>, hir_id: Option<hir::HirId>, err: &mut Diag<'_>) {
+    let Some(hir_id) = hir_id else { return };
+    let hir::Node::Ty(ty) = tcx.hir_node(hir_id) else { return };
+    let hir::TyKind::TraitObject([trait_ref, ..], ..) = ty.kind else { return };
+    let mut hir_id = hir_id;
+    while let hir::Node::Ty(ty) = tcx.parent_hir_node(hir_id) {
```

Comment:
I think you can use `parent_iter` and `take_while` rather than redoing it here.

Reply:
Something like this?

```rust
-    let mut hir_id = hir_id;
-    while let hir::Node::Ty(ty) = tcx.parent_hir_node(hir_id) {
-        hir_id = ty.hir_id;
-    }
-    if tcx.parent_hir_node(hir_id).fn_sig().is_none() {
+    let parent_ty_hir_id = tcx
+        .parent_iter(hir_id)
+        .take_while(|_id, node| if let hir::Node::Ty(_) = node { true } else { false })
+        .last()
+        .unwrap_or(hir_id);
+    if tcx.parent_hir_node(parent_ty_hir_id).fn_sig().is_none() {
```

IMO that's harder to read. I also don't have access to `parent_iter` since it's a method on `Map`, though I'm probably just missing how to access that value.

Reply:
You can access `Map` through `tcx.hir()`.

Also, I'd probably write it a bit shorter like:

```
if tcx
  .hir()
  .parent_iter(hir_id)
  .take_while(|_, node| matches!(node, hir::Node::Ty(_)))
```

Or actually, maybe just using `skip_while` to get the *next* node.

Reply:
`parent_iter` should always be preferred over repeated calls of `parent_hir_node`/`parent_hir_id` for performance reasons (see also docs of https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.parent_hir_id)

Reply:
> Or actually, maybe just using skip_while to get the next node.

nit: I don't quite see how `skip_while` would work here since the goal is to get the last element that matches a predicate. I think `take_while` + `last` seems the most straightforward.

Reply:
Yeah, and we're immediately calling `parent_hir_node` on it. So what we really want is the first node that *isn't* a `Node::Ty`. But whatever.